src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -30,124 +27,111 @@ internal static class DocumentBasedFixAllProviderHelpers
-            source: docIdToNewRootOrText,
-            produceItems: static async (tuple, callback, currentSolution, cancellationToken) =>
-            {
-                var (docId, (newRoot, _)) = tuple;
-                if (newRoot != null)
+            using var _ = await RemoteKeepAliveSession.CreateAsync(dirtySolution, cancellationToken).ConfigureAwait(false);
+
+            // Next, go and cleanup any trees we inserted. Once we clean the document, we get the text of it and insert that
+            // back into the final solution.  This way we can release both the original fixed tree, and the cleaned tree
+            // (both of which can be much more expensive than just text).
+            //
+            // Do this in parallel across all the documents that were fixed and resulted in a new tree (as opposed to new
+            // text).
+            return await ProducerConsumer<(DocumentId docId, SourceText sourceText)>.RunParallelAsync(
+                source: changedRootDocumentIds,
+                produceItems: static async (documentId, callback, args, cancellationToken) =>
                 {
-                    var cleaned = await GetCleanedDocumentAsync(
-                        currentSolution.GetRequiredDocument(docId), cancellationToken).ConfigureAwait(false);
-                    callback(cleaned);
-                }
-            },
-            consumeItems: static async (results, currentSolution, _) =>
-            {
-                // Finally, apply the cleaned documents to the solution.
-                var finalSolution = currentSolution;
-                await foreach (var (docId, cleanedText) in results)
-                    finalSolution = finalSolution.WithDocumentText(docId, cleanedText);
-
-                return finalSolution;
-            },
-            args: currentSolution,
-            cancellationToken).ConfigureAwait(false);
-
-        static async Task<(DocumentId docId, SourceText sourceText)> GetCleanedDocumentAsync(Document dirtyDocument, CancellationToken cancellationToken)
-        {
-            var cleanedDocument = await PostProcessCodeAction.Instance.PostProcessChangesAsync(dirtyDocument, cancellationToken).ConfigureAwait(false);
-            var cleanedText = await cleanedDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
-            return (dirtyDocument.Id, cleanedText);
+                    using var _ = args.progressTracker.ItemCompletedScope();
+
+                    var dirtyDocument = args.dirtySolution.GetRequiredDocument(documentId);
+                    var cleanedDocument = await PostProcessCodeAction.Instance.PostProcessChangesAsync(dirtyDocument, cancellationToken).ConfigureAwait(false);
+                    var cleanedText = await cleanedDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+                    callback((dirtyDocument.Id, cleanedText));
```

Comment:
i have a followup PR that moves this cleanup step to OOP.  so we can benefit from .net core execution wins.